daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Typing.scala

```diff
@@ -483,41 +481,82 @@ private[validation] object Typing {
 
-      impls.foreach { case (iface, impl) =>
-        val DefInterfaceSignature(requires, _, _, methods, _, _) =
-          handleLookup(ctx, interface.lookupInterface(impl.interfaceId))
-
-        requires
-          .filterNot(impls.contains)
-          .foreach(required => throw EMissingRequiredInterface(ctx, tplTcon, iface, required))
+      requires
+        .filterNot(required =>
+          interface.lookupInterfaceCoImplements(tplTcon, required).isRight ||
+            interface.lookupTemplateImplements(tplTcon, required).isRight
+        )
+        .foreach(required => throw EMissingRequiredInterface(ctx, tplTcon, ifaceTcon, required))
 
-        methods.values.foreach { (method: InterfaceMethod) =>
-          if (!impl.methods.contains(method.name))
-            throw EMissingInterfaceMethod(ctx, tplTcon, impl.interfaceId, method.name)
-        }
-        impl.methods.values.foreach { (tplMethod: TemplateImplementsMethod) =>
-          methods.get(tplMethod.name) match {
-            case None =>
-              throw EUnknownInterfaceMethod(ctx, tplTcon, impl.interfaceId, tplMethod.name)
-            case Some(method) =>
-              checkExpr(tplMethod.value, method.returnType)
-          }
+      methods.values.foreach { (method: InterfaceMethod) =>
+        if (!implMethods.exists { case (name, _) => name == method.name })
+          throw EMissingInterfaceMethod(ctx, tplTcon, ifaceTcon, method.name)
+      }
+      implMethods.foreach { case (name, value) =>
+        methods.get(name) match {
+          case None =>
+            throw EUnknownInterfaceMethod(ctx, tplTcon, ifaceTcon, name)
+          case Some(method) =>
+            checkExpr(value, method.returnType)
         }
       }
     }
 
-    // TODO (MA): https://github.com/digital-asset/daml/issues/14047
-    @nowarn("cat=unused&msg=parameter value ifaceTcon in method")
-    @nowarn("cat=unused&msg=parameter value coImpls in method")
+    private def checkIfaceImplementation(
+        tplTcon: TypeConName,
+        impl: TemplateImplements,
+    ): Unit = {
+      val ifaceTcon = impl.interfaceId
+      interface
+        .lookupInterfaceCoImplements(tplTcon, ifaceTcon)
+        .foreach(_ => throw EConflictingImplementsCoImplements(ctx, tplTcon, ifaceTcon))
+      checkGenImplementation(
+        tplTcon,
+        ifaceTcon,
+        impl.methods.values.map(TemplateImplementsMethod.unapply(_).value).toList,
+      )
+    }
+
+    private def checkIfaceImplementations(
+        tplTcon: TypeConName,
+        impls: Map[TypeConName, TemplateImplements],
+    ): Unit =
+      impls.values.foreach(checkIfaceImplementation(tplTcon, _))
+
+    private def checkIfaceCoImplementation(
+        ifaceTcon: TypeConName,
+        param: ExprVarName,
+        coImpl: InterfaceCoImplements,
+    ): Unit = {
+      val tplTcon = coImpl.templateId
+
+      // NOTE (MA): we don't need to look for conflicting TemplateImplements since
```

Comment:
the asymmetry slightly bugs me. I wonder if we might just be better off doing the check on both sides? It’s not going to have any meaningful performance impact and it’s not longer than this comment. The strongest argument I see against it is that it’s arguably shotgun ~parsing~ typechecking. Don’t feel strongly about this at all, so please go with whatever you/your team prefer :)

Reply:
I like the suggestion by @cocreature. Don't feel strongly neither though.

Reply:
I've added the symmetric check now